pkg/connector/ble/ble.go

```diff
@@ -125,10 +125,89 @@ func (c *Connection) VIN() string {
 	return c.vin
 }
 
+func VehicleLocalName(vin string) string {
+	vinBytes := []byte(vin)
+	digest := sha1.Sum(vinBytes)
+	return fmt.Sprintf("S%02xC", digest[:8])
+}
+
+func initDevice() error {
+	var err error
+	// We don't want concurrent calls to NewConnection that would defeat
+	// the point of reusing the existing BLE device. Note that this is not
+	// an issue on MacOS, but multiple calls to newDevice() on Linux leads to failures.
+	if device != nil {
+		log.Debug("Reusing existing BLE device")
+	} else {
+		log.Debug("Creating new BLE device")
+		device, err = newDevice()
+		if err != nil {
+			return fmt.Errorf("failed to find a BLE device: %s", err)
+		}
+		ble.SetDefaultDevice(device)
+	}
+	return nil
+}
+
+type Advertisement = ble.Advertisement
+
+func ScanVehicleBeacon(ctx context.Context, vin string) (Advertisement, error) {
+	mu.Lock()
+	defer mu.Unlock()
+
+	if err := initDevice(); err != nil {
+		return nil, err
+	}
+
+	a, err := scanVehicleBeacon(ctx, VehicleLocalName(vin))
+	if err != nil {
+		return nil, fmt.Errorf("ble: failed to scan for %s: %s", vin, err)
+	}
+	return a, nil
+}
+
+func scanVehicleBeacon(ctx context.Context, localName string) (Advertisement, error) {
+	var err error
+	ctx2, cancel := context.WithCancel(ctx)
+	go func() {
+		select {
+		case <-ctx.Done():
+			cancel()
+		case <-ctx2.Done():
+		}
+	}()
+
+	ch := make(chan Advertisement)
+	fn := func(a Advertisement) {
+		ln := a.LocalName()
+		if ln != localName {
+			return
+		}
+		cancel()
+		ch <- a
```

Comment:
A bit worried about creating something like https://github.com/go-ble/ble/issues/118 here. I don't believe the interface defines whether or not blocking behavior in `fn` is allowed. If `device.Scan` blocks on `fn`, then there's a deadlock because `ch` is unbuffered and nothing reads from it until `device.Scan` finishes. On the other hand, if `fn` runs asynchronously then we could leak a go routine if two BLE devices with `localName` respond with advertisements, because the second will block indefinitely while waiting for `ch` to clear.

Reply:
I would need help with this change 😅. I'm new to Go and also don't have a Mac to test.

Reply:
I based this on your changes: https://github.com/Lenart12/vehicle-command/pull/1

Tested on Linux and MacOS.